kbs/src/admin/password.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2025 by NVIDIA.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use actix_web::{http::header::Header, HttpRequest};
+use actix_web_httpauth::headers::authorization::{Authorization, Bearer};
+use argon2::{Argon2, PasswordHash, PasswordVerifier};
+use jwt_simple::{
+    claims::NoCustomClaims,
+    common::VerificationOptions,
+    prelude::{Claims, Ed25519KeyPair, EdDSAKeyPairLike, EdDSAPublicKeyLike},
+};
+use log::info;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+use crate::admin::error::*;
+use crate::admin::AdminBackend;
+
+const ADMIN_TOKEN_ISSUER: &str = "Trustee Admin";
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+#[serde(default)]
+pub struct PasswordAdminConfig {
+    pub personas: Vec<PasswordPersona>,
+    /// The number of hours that an admin token will be valid for.
+    /// After the token expires, an admin must login again and receive
+    /// a new token.
+    pub admin_token_life_hours: u64,
+    /// The key pair used for signing and validating the admin tokens.
+    /// If a path is provided, the key pair will be loaded from a file.
+    /// This be used to share auth tokens between Trustee instances and
+    /// to allow Trustee to restart with invalidating existing tokens.
+    /// The key pair should be in PEM format.
+    /// If no path is provided, a random key pair will be generated.
+    pub key_pair_path: Option<PathBuf>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct PasswordPersona {
+    pub username: String,
+    /// An argon2 pch, which includes a salt.
+    pub password_hash: String,
+}
+
+pub struct PasswordAdminBackend {
+    personas: HashMap<String, PasswordPersona>,
+    admin_token_life_hours: u64,
+    key_pair: Ed25519KeyPair,
+}
+
+impl PasswordAdminBackend {
+    pub fn new(config: PasswordAdminConfig) -> Result<Self> {
+        let mut personas = HashMap::new();
+        for p in config.personas {
+            personas.insert(p.username.clone(), p.clone());
+        }
+
+        // TODO read key pair from file
+        let key_pair = Ed25519KeyPair::generate();
```

Comment:
This is what I meant. We can't prevent users from executing the built-in random generation logic during deployment. The best approach is to decouple this logic from the rest of the code, since the keypair is a stateful attribute, and it would be better to add it to the database or some other location.

We can have this initial logic in the db setting up script.

Reply:
Yes, the reason statelessness is needed is to decouple state and computational logic in multi-replica deployment scenarios. Whether a stateless design is appropriate, or whether multi-replica deployment should be supported, is a question whether the project should address. While I personally support this direction, we still need to discuss and reach a consensus.

A compromise way is to support both, by some specific switches in the config to forward those state things to outer services.